project/app.cpp

```diff
@@ -0,0 +1,309 @@
+public:
+  explicit bench_statistics_reader(bench_statistics &stats) : _stats(stats) {}
+
+  virtual void on_input(const std::string &name) const override {
+    params local;
+    const std::string type = "INPUT";
+    local.type = type;
+    local.name = name;
+    _stats.el.push_back(local);
+    _stats.inputs.push_back(name);
+    ++_stats.number_of_inputs;
+  }
+
+  virtual void on_output(const std::string &name) const override {
+    ++_stats.number_of_outputs;
+  }
+
+  virtual void on_dff(const std::string &input,
+                      const std::string &output) const override {
+    std::string type = "DFF";
+    params local;
+    local.type = type;
+    local.gates.push_back(input);
+    local.name = output;
+    _stats.el.push_back(local);
+    ++_stats.number_of_dffs;
+  }
+
+  virtual void on_gate(const std::vector<std::string> &inputs,
+                       const std::string &output,
+                       const std::string &type) const override {
+    gate_lines.emplace_back(inputs, output, type);
+    params local;
+    local.type = type;
+    local.name = output;
+    for (int i = 0; i < inputs.size(); ++i)
+        local.gates.push_back(inputs[i]);
+    _stats.el.push_back(local);
+    _stats.number_of_lines++;
+  }
+
+  virtual void on_assign(const std::string &input,
+                         const std::string &output) const override {
+    ++_stats.number_of_lines;
+  }
+
+  bench_statistics &_stats;
+  mutable std::vector<
+      std::tuple<std::vector<std::string>, std::string, std::string>>
+      gate_lines;
+}; /* bench_statistics_reader */
+
+static void dump_statistics(FILE *f, const bench_statistics &st) {
+  fprintf(f, "inputs: %u, outputs: %u, num ddfs: %u, num lines: %u\n",
+          st.number_of_inputs, st.number_of_outputs, st.number_of_dffs,
+          st.number_of_lines);
+}
+
+void setElCoord(bench_statistics &stats, int index_number) {
+  // function that specifies coordinate for all elements
+  int maxXCoord = 0;
+
+  params &data = stats.el[index_number];
+  if (data.set_coord)
+    return;
+
+  data.setting_coord = true;
+
+  for (int i = 0; i < data.gates.size(); ++i) {
+    if (data.set_coord)
+      break;
+
+    std::string gate = data.gates.at(i);
+    int index = stats.g[gate];
+
+    if (stats.el.at(index).set_coord) {
+      maxXCoord = fmax(maxXCoord, stats.el.at(index).x);
+    } else {
+      if (stats.el.at(index).setting_coord) {
+        stats.el.at(index).setting_coord = false;
+        continue;
+      }
+      setElCoord(stats, index);
+
+      maxXCoord = fmax(maxXCoord, stats.el.at(index).x);
+    }
+  }
+  if (data.setting_coord)
+    data.setting_coord =false;
+  int y_coord = 0;
+  for (int j = 0; j < stats.el.size(); ++j) {
+    if (stats.el.at(j).x == maxXCoord + GAPS)
+      y_coord = fmax(y_coord, stats.el.at(j).y + GAPS);
+  }
+  data.x = maxXCoord + 80;
+  data.y = y_coord;
+  data.set_coord =true;
+}
+
+void setInputsCoord(bench_statistics &stats, std::string name,
```

Comment:
`stats` is not used, only `stats.el` is: change procedure parameter accordingly.